ding/policy/pc.py

```diff
@@ -0,0 +1,239 @@
+        if self._cuda:
+            collated_data = to_device(data, self._device)
+        else:
+            collated_data = data
+        observations, bfs_input_maps, bfs_output_maps = collated_data['obs'], collated_data['bfs_in'].long(), \
+                                                        collated_data['bfs_out'].long()
+        states = observations
+        bfs_input_onehot = torch.nn.functional.one_hot(bfs_input_maps, self._num_actions + 1).float()
+
+        bfs_states = torch.cat([
+            states,
+            bfs_input_onehot,
+        ], dim=-1)
+        logits = self._model(bfs_states)['logit']
+        logits = logits.flatten(0, -2)
+        labels = bfs_output_maps.flatten(0, -1)
+
+        loss = self._loss(logits, labels)
+        preds = torch.argmax(logits, dim=-1)
+        acc = torch.sum((preds == labels)) / preds.shape[0]
+
+        self._optimizer.zero_grad()
+        loss.backward()
+        self._optimizer.step()
+        pred_loss = loss.item()
+
+        cur_lr = [param_group['lr'] for param_group in self._optimizer.param_groups]
+        cur_lr = sum(cur_lr) / len(cur_lr)
+        return {'cur_lr': cur_lr, 'total_loss': pred_loss, 'acc': acc}
+
+    def _monitor_vars_learn(self):
+        return ['cur_lr', 'total_loss', 'acc']
+
+    def _init_eval(self):
+        self._eval_model = model_wrap(self._model, wrapper_name='base')
+        self._eval_model.reset()
+
+    def _forward_eval(self, data):
+        if self._cuda:
+            data = to_device(data, self._device)
+        max_len = self._max_bfs_steps
+        data_id = list(data.keys())
+        output = {}
+
+        for ii in data_id:
+            states = data[ii].unsqueeze(0)
+            bfs_input_maps = self._num_actions * torch.ones([1, self._maze_size, self._maze_size]).long()
+            if self._cuda:
+                bfs_input_maps = to_device(bfs_input_maps, self._device)
+            xy = torch.where(states[:, :, :, -1] == 1)
+            observation = (xy[1][0].item(), xy[2][0].item())
+
+            i = 0
+            while bfs_input_maps[0, observation[0], observation[1]].item() == self._num_actions and i < max_len:
+                bfs_input_onehot = torch.nn.functional.one_hot(bfs_input_maps, self._num_actions + 1).long()
+
+                bfs_states = torch.cat([
+                    states,
+                    bfs_input_onehot,
+                ], dim=-1)
+                logits = self._model(bfs_states)['logit']
+                bfs_input_maps = torch.argmax(logits, dim=-1)
+                i += 1
+            output[ii] = bfs_input_maps[0, observation[0], observation[1]]
+            if self._cuda:
+                output[ii] = {'action': to_device(output[ii], 'cpu'), 'info': {}}
+            if output[ii]['action'].item() == self._num_actions:
+                output[ii]['action'] = torch.randint(low=0, high=self._num_actions, size=[1])[0]
+        return output
+
+    def _init_collect(self) -> None:
```

Comment:
maybe you can implement all the methods related to collect with `NotImplementedError` and remove related comments